examples/ota-provider-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,240 @@
+    {
+        ESP_LOGE(TAG, "nvs_flash_init() failed: %s", esp_err_to_name(err));
+        return;
+    }
+
+    OTAProviderExample otaProvider;
+
+    CHIPDeviceManager & deviceMgr = CHIPDeviceManager::GetInstance();
+
+    CHIP_ERROR error = deviceMgr.Init(&EchoCallbacks);
+    if (error != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "device.Init() failed: %s", ErrorStr(error));
+        return;
+    }
+
+    chip::Server::GetInstance().Init();
+
+    // Initialize device attestation config
+    SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+
+    // Register handler to handle bdx messages
+    error = chip::Server::GetInstance().GetExchangeManager().RegisterUnsolicitedMessageHandlerForProtocol(chip::Protocols::BDX::Id,
+                                                                                                          &bdxServer);
+    if (error != CHIP_NO_ERROR)
+    {
+        ESP_LOGE(TAG, "RegisterUnsolicitedMessageHandler failed: %s", chip::ErrorStr(error));
+        return;
+    }
+
+    BdxOtaSenderCallbacks callbacks;
+    callbacks.onBlockQuery       = &mOnBlockQuery;
+    callbacks.onTransferComplete = &mOnTransferComplete;
+    callbacks.onTransferFailed   = &mOnTransferFailed;
+    bdxServer.SetCallbacks(callbacks);
+
+    // If OTA image is available in flash storage then set to update available
+    otaPartition = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_DATA_NVS, "test_data");
+    if (otaPartition != nullptr)
+    {
+        ESP_LOGI(TAG, "Partition found %s address:0x%x size:0x%x", otaPartition->label, otaPartition->address, otaPartition->size);
+
+        // TODO: Use the OTA image header specified in specifications
+        //       Right now we are using just image length instead of full header
+        esp_partition_read(otaPartition, 0, &otaImageLen, sizeof(otaImageLen));
+        if (otaImageLen > otaPartition->size)
+        {
+            otaImageLen = 0;
+        }
+        ESP_LOGI(TAG, "OTA image length %d bytes", otaImageLen);
+    }
+    else
+    {
+        ESP_LOGE(TAG, "OTA partition not found");
+    }
+
+    if (otaImageLen > 0)
+    {
+        otaProvider.SetQueryImageBehavior(OTAProviderExample::kRespondWithUpdateAvailable);
+        otaProvider.SetOTAFilePath(gOtaFilename);
+    }
+
+    chip::app::Clusters::OTAProvider::SetDelegate(kOtaProviderEndpoint, &otaProvider);
+
+    BitFlags<TransferControlFlags> bdxFlags;
+    bdxFlags.Set(TransferControlFlags::kReceiverDrive);
+    error = bdxServer.PrepareForTransfer(&chip::DeviceLayer::SystemLayer(), chip::bdx::TransferRole::kSender, bdxFlags,
+                                         kMaxBdxBlockSize, kBdxTimeout, kBdxPollFreq);
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "Failed to init BDX server: %s", chip::ErrorStr(error));
+        return;
+    }
+
+    // Run the UI Loop
+    while (true)
+    {
+        vTaskDelay(50 / portTICK_PERIOD_MS);
+    }
```

Comment:
Why is this while loop required?

Reply:
This example is inherited from `examples/temperature-measurement-app/esp32`, all the ESP32 examples have this loop. Although, I tried removing it and device was not advertising so, I think those are needed.